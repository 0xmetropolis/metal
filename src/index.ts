#!/usr/bin/env node

import * as yargs from 'yargs';

// prettier-ignore
yargs
  .commandDir('commands')
  .demandCommand(1)
  .help()
  .version()
  .argv;
