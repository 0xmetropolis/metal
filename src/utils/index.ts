import chalk from 'chalk';
import { Abi, FoundryConfig, HexString } from 'index';
import { emojify } from 'node-emoji';
import { exec, execSync } from 'node:child_process';
import { UUID } from 'node:crypto';
import { readFileSync, readdirSync } from 'node:fs';
import { getOutPath, isSparseModeEnabled } from './foundry';
import fetch from 'node-fetch';

export const logError = (...s: string[]) =>
  console.log('\n\n\n' + emojify(chalk.bold.red(s.join('\n'))));
export const logInfo = (s: string) => console.log(emojify(chalk.bold(s)));
export const logDebug = (s: string | any) =>
  !!process.env.DEBUG &&
  console.log(chalk.yellowBright(typeof s === 'object' ? JSON.stringify(s?.message ?? s) : s));
export const logWarn = (...s: string[]) =>
  console.warn(emojify(chalk.bold.yellow(s.map(str => '⚠️ ' + str + ' ⚠️').join('\n'))));
export const logDetail = (s: string) => console.log(emojify(chalk.dim(s)));

export const exit = (...message: string[]) => {
  logError.call(this, message.join('\n'));
  process.exit(1);
};

export const getFlagValueFromArgv = (flag: string): string | undefined => {
  const flagIndex = process.argv.findIndex(arg => arg === flag);
  if (flagIndex === -1) return undefined;
  else return process.argv[flagIndex + 1];
};

export const openInBrowser = (url: string) => {
  const startScript =
    process.platform == 'darwin' ? 'open' : process.platform == 'win32' ? 'start' : 'xdg-open';
  exec(`${startScript} "${url}"`);
};

export const getChainId = async (rpcUrl: string) => {
  try {
    const request = await fetch(rpcUrl, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
      },
      body: JSON.stringify({ jsonrpc: '2.0', method: 'eth_chainId', params: [], id: 0 }),
    });

    const response: { jsonrpc: '2.0'; id: 0; result: HexString } = await request.json();

    return Number(response.result); // convert hex to number
  } catch (e: any) {
    logDebug(e);
    exit('Error fetching chainId from RPC endpoint');
  }
};

export const getTransactionByHash = async (rpcUrl: string, txHash: string) => {
  try {
    if (typeof txHash !== 'string' || txHash.length !== 66)
      throw new Error(`Invalid transaction hash: ${txHash}`);

    const request = await fetch(rpcUrl, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
      },
      body: JSON.stringify({
        jsonrpc: '2.0',
        method: 'eth_getTransactionByHash',
        params: [txHash],
        id: 0,
      }),
    });

    const response: {
      jsonrpc: '2.0';
      id: number;
      result: {
        blockHash: HexString | null;
        blockNumber: HexString | null;
        from: HexString;
        gas: HexString;
        gasPrice: HexString;
        hash: HexString;
        input: HexString;
        nonce: HexString;
        to: HexString;
        transactionIndex: HexString;
        value: HexString;
        v: HexString;
        r: HexString;
        s: HexString;
      } | null;
    } = await request.json();

    return response.result;
  } catch (e: any) {
    logDebug(e);
    exit(`Error fetching transaction: ${txHash} from RPC endpoint`);
  }
};

export const isTxConfirmed = async (rpcUrl: string, txHash: string) => {
  const response = await getTransactionByHash(rpcUrl, txHash);
  if (response === null || response.blockNumber === null) return false;

  return true;
};

export const getConfigFromTenderlyRpc = (rpcOverride: string): { rpcUrl: string; id: UUID } => ({
  rpcUrl: rpcOverride,
  id: rpcOverride.split('/').at(-1) as UUID,
});

export const replaceFlagValues = ({
  args,
  flags,
  replaceWith: replaceWith,
}: {
  args: string[];
  flags: string[];
  replaceWith: string;
}) => {
  const flagIndex = args.findIndex(arg => flags.some(alias => alias === arg));

  if (flagIndex === -1) return args;
  else return args.map((arg, index) => (index === flagIndex + 1 ? replaceWith : arg));
};

// @dev returns kv pairs of file path and source code
export const loadSolidityFiles = (pathToSolc: string[]): { [filePath: string]: string } => {
  const sourceCode = pathToSolc.reduce<{ [contractFilePath: string]: string }>(
    (sourceCodeAcc, path) => {
      try {
        const code = readFileSync(path, { encoding: 'utf-8' });
        return { ...sourceCodeAcc, [path]: code };
      } catch (e: any) {
        logDebug(e);
        exit(`Could not find Solidity source file: ${path}`);
      }
    },
    {},
  );

  return sourceCode;
};

/**
 * @dev based on the absolute path to the solidity files in `src/`, load the metadata abi files in `out/`
 * @notice a `fullyQualifiedContractName` is the path to the solidity file concatenated with ":CONTRACT_NAME" (e.g: `contracts/MyScript.s.sol:MyCoolScript`)
 * @param pathsToSolidityFiles paths to all solidity files
 */
export const loadSolidityABIs = (foundryConfig: FoundryConfig, pathsToSolidityFiles: string[]) => {
  if (isSparseModeEnabled(foundryConfig))
    exit(
      'sparse_mode is enabled in foundry.toml, preventing artifact generation.',
      'Please disable sparse_mode and try again.',
    );

  // find the correct `out` path where the metadata files go by checking the foundry.toml
  const outPath = getOutPath(foundryConfig);

  let allABIPaths: string[] = [];
  try {
    // recursively builds an array any *.sol directories
    //   (will match `out/MyContract.sol/` and `out/long/dependency/path/MyContract.sol/)`
    allABIPaths = execSync(`find ${outPath} -type d -name "**.sol"`).toString().split('\n');
  } catch (e) {
    logDebug(e);
    exit(`Error finding build metadata in the ${outPath} directory`);
  }

  // collect all abis from the metadata files
  const abis = pathsToSolidityFiles.reduce<{ [fullyQualifiedContractName: string]: string }>(
    (abiAcc, pathToSolidityFile) => {
      // get the file name by grabbing the last item in the path
      const fileName = pathToSolidityFile.split('/').at(-1); // e.g: /Users/.../contracts/MyContract.sol -> MyContract.sol

      // the path to all the contract ABIs for a given solidity file
      const fullMetadataFilDirPath = allABIPaths.find((path: string) => {
        // the directory name is the same as the solidity file name
        const fileMetadataDir = path.split('/').at(-1).toLowerCase();

        return fileMetadataDir === fileName.toLowerCase();
      });

      if (!fullMetadataFilDirPath) {
        logError(`Could not find metadata for ${fileName}`);
        // do not try and load metadata if the metadata dir
        return abiAcc;
      }

      let files: string[] = [];
      try {
        // all the child files in a sol metadata dir are contract ABIs: e.g: ['MyContract.json', 'MyContract2.json']
        files = readdirSync(fullMetadataFilDirPath, { encoding: 'utf-8' });
      } catch (e: any) {
        logDebug(e);
        logError(`Could not find metadata for ${fullMetadataFilDirPath}`);
      }

      // for each pathToSolc, load all child files in the `out/MyContract.sol/` directory.
      const abis = files.reduce<[string, Abi][]>((abiAcc, file) => {
        try {
          const metadataPath = `${fullMetadataFilDirPath}/${file}`;
          // JSON.parse each file and return only the .abi member.
          const abi = JSON.parse(readFileSync(metadataPath, 'utf-8')).abi;
          if (!abi) throw new Error(`ABI not found for ${fullMetadataFilDirPath}/${file}`);

          const contractName = file.replace('.json', '');
          const fullyQualifiedContractName = `${pathToSolidityFile}:${contractName}`;
          return [...abiAcc, [fullyQualifiedContractName, abi]];
        } catch (e: any) {
          logDebug(e);
          // handle error if any of the json is invalid / missing .abi
          logError(`Could not find ABI for ${file}`);
        }
      }, []);

      // return the abiAcc with the new abis
      return { ...abiAcc, ...Object.fromEntries(abis) };
    },
    {},
  );

  return abis;
};
