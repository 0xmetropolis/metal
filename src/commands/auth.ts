import { Arguments, Options } from 'yargs';
import { logDebug, logError, logInfo } from '../utils';
import { authenticateViaPCKEFlow, checkAuthentication } from '../utils/auth';
import { sendCliCommandAnalytics } from '../utils/analytics';

export const command = 'auth';
export const description = `Authenticate with Metal`;

export type Params = {};
export type HandlerInput = Arguments & Params;

export const builder: { [key: string]: Options } = {
  force: {
    type: 'boolean',
    required: false,
    description: 'Force re-authentication even if already authenticated',
  },
};

/**
 * @dev entry point for the auth command
 * @param force - the user can force re-authentication, even if their token is valid
 */
export const handler = async ({ force }: HandlerInput) => {
  // check if the user is already authenticated (includes an expiry check and a refresh - if necessary)
  const auth = await checkAuthentication();

  // debug for development
  if (auth.status === 'authenticated') logDebug(JSON.stringify(auth, null, 2));

  // if the user is already authenticated, bail early
  if (auth.status === 'authenticated' && !force) {
    logInfo('Already authenticated ðŸŽ‰\n\nðŸ’¡ Use the `--force` flag to re-authenticate');

    return;
  }

  /**
   * @dev we use the "PCKE Authorization Flow" to authenticate with Auth0
   *   See this diagram for a visual representation of the flow:
   *   https://auth0.com/docs/get-started/authentication-and-authorization-flow/authorization-code-flow-with-proof-key-for-code-exchange-pkce#how-it-works
   */
  try {
    const { nickname } = await authenticateViaPCKEFlow();

    logInfo(`Successfully authenticated as ${nickname} ðŸŽ‰`);
  } catch (err) {
    logError(err);
    logError('Authentication failed');
  }

  await sendCliCommandAnalytics('auth');

  // hard exit to clean up any open sockets
  process.exit(0);
};
