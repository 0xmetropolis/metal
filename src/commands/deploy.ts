import { DeploymentRequestParams, Network } from 'index';
import { type Arguments, type Options } from 'yargs';
import {
  FORGE_FORK_ALIASES,
  PREVIEW_WEB_URL,
  SUPPORTED_CHAINS,
  doNotCommunicateWithPreviewService,
} from '../constants';
import { exit, getFlagValueFromArgv, logDebug, logInfo, openInBrowser } from '../utils';
import {
  getContractMetadata,
  getScriptDependencies,
  getScriptMetadata,
  loadFoundryConfig,
  normalizeForgeScriptPath,
  runForgeScript,
} from '../utils/foundry';
import { getRepoMetadata } from '../utils/git';
import { ChainConfig, fetchChainConfig, uploadDeploymentData } from '../utils/preview-service';
import { getCLIVersion } from '../utils/version';
import inquirer = require('inquirer');

export const command = 'deploy';
export const description = `Run your deployments against a live network and generate a Metropolis preview`;

export type Params = { 'chain-id': number };
export type HandlerInput = Arguments & Params;

export const builder: { [key: string]: Options } = {
  'chain-id': {
    type: 'number',
    required: true,
    description: 'The chain id of the network',
  },
  broadcast: {
    type: 'boolean',
    required: true,
    description: 'Must broadcast transactions to the network',
  },
};

function validateInputs({ _: [, scriptPath], 'chain-id': chainId }: HandlerInput) {
  if (!scriptPath || !scriptPath.includes('.sol'))
    exit('You must specify a solidity script to preview');

  const rpcSpecified = FORGE_FORK_ALIASES.some(alias => process.argv.includes(alias));

  // if the rpc is specified, we don't need to validate the chain id
  if (!SUPPORTED_CHAINS.includes(chainId) && !rpcSpecified)
    exit(`Chain Id ${chainId} is not supported`);
}

// @dev pulls any args from process.argv and replaces any fork-url aliases with the preview-service's fork url
export const configureForgeScriptInputs = ({ rpcUrl }: { rpcUrl: string }): string[] => {
  // pull anything after `metro preview <path>` as forge arguments
  let forgeArguments = process.argv.slice(3);

  // rewrap function signatures in quotes, ex: --sig "run()"
  forgeArguments = forgeArguments.map(arg =>
    arg.includes('(') && arg.includes(')') ? `"${arg}"` : arg,
  );

  forgeArguments.push('--rpc-url', rpcUrl);

  if (!forgeArguments.includes('--slow')) forgeArguments.push('--slow');

  return forgeArguments;
};

const getChainConfig = async (chainId: Network): Promise<Partial<ChainConfig>> => {
  const rpcOverrideFlagIdx = process.argv.findIndex(arg => FORGE_FORK_ALIASES.includes(arg));
  const userHasSpecifiedRPC = rpcOverrideFlagIdx !== -1;

  const emptyConfig: Partial<ChainConfig> = {
    rpcUrl: undefined,
    label: undefined,
    chainId,
    etherscanUrl: undefined,
  };

  return doNotCommunicateWithPreviewService
    ? emptyConfig
    : !!userHasSpecifiedRPC
    ? { ...emptyConfig, rpcUrl: getFlagValueFromArgv(process.argv[rpcOverrideFlagIdx + 1]) }
    : await fetchChainConfig(chainId);
};

// @dev entry point for the preview command
export const handler = async (yargs: HandlerInput) => {
  validateInputs(yargs);

  // @dev arg 0 is the command name: e.g: `preview`
  const {
    _: [, forgeScriptPath],
    'chain-id': chainId,
  } = yargs;

  const { rpcUrl, label } = await getChainConfig(chainId);

  logDebug(`Loading foundry.toml...`);
  const foundryConfig = loadFoundryConfig();

  const foundryArguments = configureForgeScriptInputs({
    rpcUrl,
  });

  await inquirer
    .prompt({
      type: 'confirm',
      name: 'confirm',
      message: `You are about to sign and send transactions to ${
        label ? label : 'chain with id: ' + chainId
      }❗️\n\nAre you sure you want to continue?`,
    })
    .then(({ confirm }) => !confirm && exit('Aborting deployment'));

  logInfo(`Running Forge Script at ${forgeScriptPath}...`);
  await runForgeScript(foundryArguments);
  logInfo(`Forge deployment script ran successfully!`);

  const normalizedScriptPath = normalizeForgeScriptPath(forgeScriptPath);
  const solidityFilePaths = [
    normalizedScriptPath,
    ...getScriptDependencies(foundryConfig, normalizedScriptPath),
  ];

  logDebug(`Getting repo metadata...`);
  const repoMetadata = getRepoMetadata(solidityFilePaths);
  const cliVersion = getCLIVersion();

  logInfo(`Getting transaction data...`);
  const scriptMetadata = getScriptMetadata(foundryConfig, chainId, forgeScriptPath);

  logDebug(`Getting contract metadata...`);
  const contractMetadata = getContractMetadata(
    foundryConfig,
    scriptMetadata.broadcastArtifacts,
    solidityFilePaths,
  );

  const payload: DeploymentRequestParams = {
    cliVersion,
    chainId,
    repoMetadata,
    scriptMetadata,
    contractMetadata,
  };

  const deploymentId = doNotCommunicateWithPreviewService
    ? undefined
    : await uploadDeploymentData(payload);
  const metropoliswebUrl = `${PREVIEW_WEB_URL}/preview/${deploymentId}`;

  logInfo(`Deployment successful! 🎉\n\n`);
  logInfo(`
                             ^
                _______     ^^^
               |xxxxxxx|  _^^^^^_
               |xxxxxxx| | [][][]|
            ______xxxxx| |[][][] |
           |++++++|xxxx| | [][][]|      METROPOLIS
           |++++++|xxxx| |[][][] |
           |++++++|_________ [][]|
           |++++++|=|=|=|=|=| [] |
           |++++++|=|=|=|=|=|[][]|
___________|++HH++|  _HHHH__|   _________   _________  _________

${metropoliswebUrl}
__________________  ___________    __________________    ____________
  `);

  openInBrowser(metropoliswebUrl);
};
