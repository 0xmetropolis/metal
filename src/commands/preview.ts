import { PreviewRequestParams } from 'index';
import { type Arguments, type Options } from 'yargs';
import { METRO_DEPLOY_URL } from '../constants';
import { loadSolidityFiles, logError, logInfo, logWarn } from '../utils';
import {
  getBroadcastArtifacts,
  getScriptDependencies,
  loadFoundryConfig,
  runForgeScript,
} from '../utils/foundry';
import assert = require('node:assert');

export const command = 'preview';
export const description = `Generate preview of transactions from your Forge script`;
const FORGE_FORK_ALIASES = ['--fork-url', '-f', '--rpc-url'];

export type Params = { path: string; broadcast: boolean };
export const builder: { [key: string]: Options } = {
  broadcast: {
    type: 'boolean',
    required: true,
    description: 'Send the transaction to the metropolis RPC',
  },
};

// @dev pulls any args from process.argv and replaces any fork-url aliases with the METRO_DEPLOY_URL
export const configureForgeScriptInputs = (): string[] => {
  // pull anything after `metro preview <path>` as forge arguments
  const initialArgs = process.argv.slice(3);
  const rpcIndex = initialArgs.findIndex(arg => FORGE_FORK_ALIASES.some(alias => alias === arg));
  const userHasSpecifiedRPC = rpcIndex !== -1;

  if (userHasSpecifiedRPC)
    logWarn(
      'You have specified a custom RPC',
      'This will be ignored and transactions will be sent to the Metropolis RPC',
    );

  const formattedArgs = userHasSpecifiedRPC
    ? initialArgs.map((arg, index) => (index === rpcIndex + 1 ? METRO_DEPLOY_URL : arg))
    : [...initialArgs, '--rpc-url', METRO_DEPLOY_URL];

  return formattedArgs;
};

function devModeSanityChecks({ sourceCode, broadcastArtifacts }: PreviewRequestParams) {
  assert(Object.values(sourceCode).length > 0 && Object.values(sourceCode).every(Boolean));
  assert(broadcastArtifacts.transactions.length > 0);
  logInfo(`DEV: checks pass âœ…`);
}

export const sendDataToPreviewService = async (payload: PreviewRequestParams): Promise<string> => {
  try {
    const response = await fetch(`${METRO_DEPLOY_URL}/preview`, {
      headers: {
        'Content-Type': 'application/json',
      },
      method: 'POST',
      body: JSON.stringify(payload),
    });

    if (response.status !== 200) {
      logError('Error received from preview service:');
      logError('Status Code: ' + response.status);
      logError('Status Text: ' + response.statusText);
      process.exit(1);
    }

    const { previewURL }: { previewURL: string } = await response.json();

    return previewURL;
  } catch (e: any) {
    logError('Error connecting to preview service');
    logError(e.message);
    process.exit(1);
  }
};

// @dev entry point for the preview command
export const handler = async ({ _: [, forgeScriptPath] }: Arguments) => {
  logInfo(`Loading foundry.toml...`);
  const foundryConfig = loadFoundryConfig();

  logInfo(`Running Forge Script at ${forgeScriptPath}...`);

  const foundryArguments = configureForgeScriptInputs();
  await runForgeScript(foundryArguments);

  logInfo(`Forge deployment script ran successfully!`);

  logInfo(`Retreiving Solidity source code...`);
  const dependencyList = getScriptDependencies(foundryConfig, forgeScriptPath);
  const solidityFiles = [forgeScriptPath, ...dependencyList];
  const sourceCode = loadSolidityFiles(solidityFiles);

  logInfo(`Getting transactions...`);
  const broadcastArtifacts = await getBroadcastArtifacts(foundryConfig, forgeScriptPath);

  const payload = {
    broadcastArtifacts,
    sourceCode,
  };
  devModeSanityChecks(payload);

  const previewURL = await sendDataToPreviewService(payload);

  logInfo(`Preview simulation successful! ðŸŽ‰\n\n`);
  logInfo(`View preview: ${previewURL}`);
  logInfo(`
                             ^
                _______     ^^^
               |xxxxxxx|  _^^^^^_
               |xxxxxxx| | [][]  |
            ______xxxxx| |[][][] |
           |++++++|xxxx| | [][][]|      METROPOLIS
           |++++++|xxxx| |[][][] |
           |++++++|_________ [][]|
           |++++++|=|=|=|=|=| [] |
           |++++++|=|=|=|=|=|[][]|
___________|++HH++|  _HHHH__|   _________   _________  _________

${previewURL}
__________________  ___________    __________________    ____________
  `);
};
