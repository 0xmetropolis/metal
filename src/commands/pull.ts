import AdmZip from 'adm-zip';
import { type Arguments, type Options } from 'yargs';
import { exit, logInfo } from '../utils';
import { getMetalCodegenFromArtifactBundle, logMetalDirStructure } from '../utils/codegen';
import { getDeploymentArtifacts } from '../utils/preview-service';
import { decompressArtifactZip, tryLoadPreviousAddresses } from '../utils/pull';
import { writeCodegenToDisk } from '../utils/pull';

export const command = 'pull';
export const description = `Pull artifacts from `;

export type Params = {};
export type HandlerInput = Arguments & Params;

export const builder: { [key: string]: Options } = {};

const extractArguments = (yargs: Arguments) => {
  const {
    _: [, deploymentId],
  } = yargs;

  return { deploymentId };
};

const validateInputs = (yargs: Arguments) => {
  const { deploymentId } = extractArguments(yargs);
  if (!deploymentId) throw exit('You must specify a deployment id to pull!');

  return { deploymentId };
};

/**
 * @dev entry point for the pull command
 */
export const handler = async (yargs: HandlerInput) => {
  // validate there is an id
  validateInputs(yargs);

  const { deploymentId } = extractArguments(yargs);

  // try and load any previous addresses from a previous config file
  const previouslyDeployedAddresses = await tryLoadPreviousAddresses();

  // fetch the artifacts zip as a buffer from the backend
  const rawZip = await getDeploymentArtifacts(deploymentId, previouslyDeployedAddresses);

  // unzip the file in memory
  const zipInstance = new AdmZip(rawZip);

  // decompress to an in-memory object
  const artifactBundle = decompressArtifactZip(zipInstance);

  // generate some nice type-safe codegen
  const metalCodegen = getMetalCodegenFromArtifactBundle(artifactBundle);

  // save the codegen as .ts files
  writeCodegenToDisk(metalCodegen);

  logInfo('Wrote deployment artifacts to the `metal/` directory ðŸŽ‰\n');
  logMetalDirStructure();
};
