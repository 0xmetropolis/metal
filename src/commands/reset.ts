import { rm } from 'node:fs/promises';
import { Arguments, Options } from 'yargs';
import { ID_TOKEN_FILE } from '../constants';
import { logInfo } from '../utils';
import { deleteFromFilestore, getFilestoreDir, isInFilestore } from '../utils/filestore';

export const command = 'reset';
// this command is hidden from the help menu
export const description = false;

export type Params = {};
export type HandlerInput = Arguments & Params;

export const builder: { [key: string]: Options } = {
  auth: {
    type: 'boolean',
    required: false,
    description: 'Reset only the authentication cache',
  },
};

/**
 * @dev entry point for the reset command
 * @param auth the user can choose to reset only authentication
 */
export const handler = async ({ auth }: HandlerInput) => {
  if (auth) {
    logInfo('Resetting authentication cache...');

    if (isInFilestore(ID_TOKEN_FILE)) deleteFromFilestore(ID_TOKEN_FILE);
    return;
  }

  logInfo('Resetting entire cache...');
  await rm(getFilestoreDir(), { recursive: true });
};
