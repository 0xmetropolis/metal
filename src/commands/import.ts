import { type Arguments, type Options } from 'yargs';
import { PREVIEW_WEB_URL, doNotCommunicateWithPreviewService } from '../constants';
import { DeploymentRequestParams, ScriptMetadata } from '../types';
import {
  getFlagValueFromArgv,
  logInfo,
  openInBrowser,
  printPreviewLinkWithASCIIArt,
} from '../utils';
import {
  getContractMetadata,
  getScriptDependencies,
  loadBroadcastArtifacts,
  loadFoundryConfig,
  resolveTargetContract,
} from '../utils/foundry';
import { checkoutToCommit, getGitMetadata, getRepoMetadata } from '../utils/git';
import {
  buildProject,
  checkRepoForUncommittedChanges,
  ensureBroadcastArtifactValidityAndContinue,
  ensureRepoIsOnCorrectCommit,
  findScriptPath,
  promptForBroadcastArtifact,
} from '../utils/import';
import { fetchChainConfig, uploadDeploymentData } from '../utils/preview-service';
import { getCLIVersion } from '../utils/version';
import { checkAuthentication } from '../utils/auth';

export const command = 'import';
export const description = `Import past deployments into a Metal preview`;

export type Params = {
  'build-options': string[];
};
export type HandlerInput = Arguments & Params;

export const builder: { [key: string]: Options } = {
  'build-options': {
    type: 'array',
    required: false,
    description: 'Parameters to be passed to the `forge build` command',
  },
};

/**
 * @dev entry point for the import command
 */
export const handler = async (yargs: HandlerInput) => {
  checkRepoForUncommittedChanges();

  const authenticationStatus = await checkAuthentication();

  const foundryConfig = loadFoundryConfig();

  const { selectedScript, artifactPath, selectedChainId } = await promptForBroadcastArtifact(
    foundryConfig,
  );

  const broadcastArtifacts = loadBroadcastArtifacts(artifactPath);

  const chainConfig = await fetchChainConfig(selectedChainId);

  const continueWithBroadcastFile = await ensureBroadcastArtifactValidityAndContinue(
    artifactPath,
    broadcastArtifacts,
    chainConfig,
  );

  // if the user doesn't want to continue with the broadcast file or it's invalid, re-run the import command
  // @alert, I am unsure if yargs injects runtime vars that could break a recursive call to `handler`
  //  let's watch for bugs around retries.
  if (!continueWithBroadcastFile) {
    handler(yargs);
    return;
  }

  // ensure the directory is on the correct commit - this may return the current commit to revert to after the import
  let returnToCommitAfterImport: string | null = await ensureRepoIsOnCorrectCommit(
    broadcastArtifacts,
  );

  // run forge build now that we're on the correct commit sha
  await buildProject();

  // search for the script file in the current directory (if necessary, prompt the user to select one)
  const pathToScript = await findScriptPath(foundryConfig, selectedScript);
  // get an array of all the solidity files that need to be uploaded
  const solidityFilePaths = [pathToScript, ...getScriptDependencies(foundryConfig, pathToScript)];
  const repoMetadata = getRepoMetadata(solidityFilePaths);

  const targetContract = resolveTargetContract(selectedScript);
  const functionName = getFlagValueFromArgv('-s') || getFlagValueFromArgv('--sig') || 'run()';
  const scriptGitMetadata = getGitMetadata(selectedScript);

  const scriptMetadata: ScriptMetadata = {
    scriptName: targetContract,
    functionName,
    filePath: selectedScript,
    broadcastArtifacts,
    ...scriptGitMetadata,
  };

  const contractMetadata = getContractMetadata(
    foundryConfig,
    scriptMetadata.broadcastArtifacts,
    solidityFilePaths,
  );

  const payload: DeploymentRequestParams = {
    prompt: 'import',
    cliVersion: getCLIVersion(),
    chainId: selectedChainId,
    repoMetadata,
    scriptMetadata,
    contractMetadata,
  };

  logInfo(`Uploading repo metadata...`);

  const authToken = authenticationStatus.isAuthenticated
    ? authenticationStatus.access_token
    : undefined;

  const deploymentId = doNotCommunicateWithPreviewService
    ? undefined
    : await uploadDeploymentData(payload, authToken);
  const metalWebUrl = `${PREVIEW_WEB_URL}/preview/${deploymentId}`;

  logInfo(`Upload Successful! ðŸŽ‰\n\n`);
  printPreviewLinkWithASCIIArt(metalWebUrl);

  openInBrowser(metalWebUrl);

  if (returnToCommitAfterImport) checkoutToCommit('-', { silent: true });
};
