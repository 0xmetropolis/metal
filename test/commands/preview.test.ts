import { handler as preview } from '../../src/commands/preview';
import { runWithAnswers, ENTER } from '../utils';
import * as utils from '../../src/utils';

// silence outputs
jest.mock('inquirer/lib/utils/screen-manager');
jest.mock('../../src/utils');

const logInfo = jest.spyOn(utils, 'logInfo');

afterEach(() => {
  logInfo.mockReset();
});

describe('preview', () => {
  it('accepts a preview command with a path', async () => {
    await runWithAnswers(() => preview({ _: ['preview', 'fake/path.sol'], $0: 'preview' }), []);
    expect(logInfo).toHaveBeenCalledTimes(1);
    expect(logInfo.mock.calls[0][0]).toMatchSnapshot();
  });
});
