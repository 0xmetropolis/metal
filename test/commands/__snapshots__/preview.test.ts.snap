// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`preview says hello with the right name, without asking it if provided 1`] = `"Running Forge Script at fake/path.sol..."`;
